### PR TITLE
Use urlEncode on object keys.

### PR DESCRIPTION
**Pull request by kantp**
I ran into an error trying to upload an object to S3 with a space in the name, which has to be url encoded. I added urlEncode to the relevant parts of the library in order to fix that.

### CODE REVIEW

File: Aws/S3/Commands/CopyObject.hs [hunk 1]
```diff
@@ -55,12 +56,12 @@ instance SignQuery CopyObject where
     signQuery CopyObject {..} = s3SignQuery S3Query {
                                  s3QMethod = Put
                                , s3QBucket = Just $ T.encodeUtf8 coBucket
-                               , s3QObject = Just $ T.encodeUtf8 coObjectName
+                               , s3QObject = Just . HTTP.urlEncode False $ T.encodeUtf8 coObjectName
```
Comment (aristidb) [inline Aws/S3/Commands/CopyObject.hs]: Can you do the HTTP.urlEncode in s3SignQuery instead? Then it should not be necessary to do it in 5 different places.

File: Aws/S3/Commands/CopyObject.hs [hunk 2]
```diff
@@ -55,12 +56,12 @@ instance SignQuery CopyObject where
...
                                , s3QSubresources = []
                                , s3QQuery = []
                                , s3QContentType = coContentType
                                , s3QContentMd5 = Nothing
-                               , s3QAmzHeaders = map (second T.encodeUtf8) $ catMaybes [
+                               , s3QAmzHeaders = map (second (HTTP.urlEncode False . T.encodeUtf8)) $ catMaybes [
```
Comment (aristidb) [inline Aws/S3/Commands/CopyObject.hs]: Did you verify that this is necessary? URL Encoding HTTP headers seems weird to me.

**Comment by aristidb** [comment]
Sorry for not looking at this quicker. :)

File: Aws/S3/Commands/CopyObject.hs (hunk 1 above)
Comment (kantp) [inline Aws/S3/Commands/CopyObject.hs]: That seems like a good idea, I'll do that :)

File: Aws/S3/Commands/CopyObject.hs (hunk 2 above)
Comment (kantp) [inline Aws/S3/Commands/CopyObject.hs]: Sorry, I think that was a slip.

**Comment by aristidb** [comment]
Looks good now, I'll squash the commits.

**Comment by aristidb** [comment]
Merged manually

**Comment by kantp** [comment]
Thanks!

**Pull request closed without merging**